Whitelist/Blacklist mail domain and TLD allowed

carlitorweb opened the pull request:
### Summary of Changes
This PR adds a restriction in case the user registration is needed, only for a group of mail domain and/or TLD, or block for a group of mail domain and/or TLD.

### Testing Instructions for the option "Whitelist"
1. Go to the configs in the user manager
2. Select a _Whitelist_ option in the field **Email/TLD Restriction**. Fill the field **Email/TLD List** and save.
3. Go to the frontend/backend, and try to register/edit a user with a different domain/TLD than the field **Email/TLD List**. Also, check this is valid for the **Edit Account** option in the top right corner of the admin area. 
4. Go to the frontend/backend, and try to register/edit a user with a domain/TLD which is on the list **Email/TLD List**. Also, check this is valid for the "Edit Account" option in the top right corner of the admin area. 

### Expected result
An error message will show up, saying, "Your mail domain xxx is not allowed", refusing registration in case you try to register with a different domain/TLD. Or the new registration is accepted otherwise.

### Testing Instructions for the option "Blacklist"
1. Go to the configs in the user manager
2. Select a _Blacklist_ option in the field **Email/TLD Restriction**. Fill the field **Email/TLD List** and save.
3. Go to the frontend/backend, and try to register/edit a user with a different domain/TLD than the field **Email/TLD List** . Also, check this is valid for the **Edit Account** option in the top right corner of the admin area. 
3. Go to the frontend/backend, and try to register/edit a user with a domain/TLD which is on the list **Email/TLD List**. Also, check this is valid for the **Edit Account** option in the top right corner of the admin area. 

### Expected result
An error message will show up, saying, "Your mail domain xxx is not allowed", refusing registration in case you try to register with a domain/TLD, which it is within of the field **Email/TLD List**. Or the new registration is accepted otherwise.

### Documentation Changes Required
New com_users config added.

carlitorweb committed:
Unique email domain allowed

This PR add a restriction in case the user registration is needed, only
from a single mail domain.

##How to test
1. Go to the configs in the user manager
2. Fill the field "Unique mail domain", and save
3. Go to the frontend, and try to register with a different domain than
the field "Unique mail domain". A error message will show up, saying,
"Your mail domain xxx is not allowed".

SharkyKZ commented:
If this feature is approved, it should be improved to accept multiple domains and to set whether domains are allowed or banned.

brianteeman commented:
yes it should be a - a whitelist/blacklist feature

alikon commented:
is this a core feature ?

Bakual commented:
> is this a core feature ?

Imho no. I'm quite sure there exist already 3rd party plugins which do that fine and the use case is quite limited.

brianteeman commented:
I looked and didnt see any

Enhanced to have a list of allowed and disallowed domains could be quite useful

ot2sen commented:
Could be useful in core. But there are free extensions doing feature like this. For example:
https://extensions.joomla.org/extensions/extension/access-a-security/site-access/domainrestriction/

brianteeman commented:
i missed that

carlitorweb commented:
I'm going to implement the Whitelis/Blacklist. In my opinion, is useful and necessary this featured inside core.

carlitorweb committed:
Whitelist/Blacklist mail domain

carlitorweb commented:
This featured, It is normal in all kinds of scenarios, such as user registration. I do not think you should **install another extension**, to have it enabled

carlitorweb committed:
[mod_articles_news]  Option Intro/Full Images of the article

Option to use or not, the intro image field or the full image field of
the article

carlitorweb committed:
Revert "[mod_articles_news]  Option Intro/Full Images of the article"

This reverts commit 6d5ae14ddc9257a4b6ad9c9669ccd7df7a71072e.

carlitorweb commented:
Sorry for this, did wrong here

stutteringp0et commented:
The extension @ot2sen mentioned is mine - I originally wrote it for a school, to limit registrations to employees using a valid school domain.  Later I expanded it on user feature requests.

I find that most people use it to blacklist domains, but it can white/black list TLD, Domain, and full addresses.  So you could limit registrations to only .CU domains if you wanted....

Quy commented on components/com_users/controllers/registration.php:
```diff
@@ -105,7 +105,7 @@ public function activate()
 	 * @since   1.6
 	 */
-	public function register()
+	public function register() 
```
Revert change.

Quy commented on components/com_users/models/registration.php:
```diff
@@ -422,6 +422,17 @@ public function register($temp)
 		$sendpassword = $params->get('sendpassword', 1);
+		$whiteListMailDomain = explode("\r\n", $params->get('whiteListMailDomain'));
+		$blackListMailDomain = explode("\r\n", $params->get('blackListMailDomain'));
+		$userMailDomain = explode('@',$data['email']);
```
Add space after comma.

Quy commented on administrator/language/en-GB/en-GB.com_users.ini:
```diff
@@ -17,6 +17,8 @@ COM_USERS_CATEGORIES_TITLE="User Notes: Categories"
 COM_USERS_CATEGORY_HEADING="Category"
 COM_USERS_CONFIG_FIELD_ALLOWREGISTRATION_DESC="If set to Yes, new Users are allowed to self-register."
 COM_USERS_CONFIG_FIELD_ALLOWREGISTRATION_LABEL="Allow User Registration"
+COM_USERS_CONFIG_FIELD_BLACKLIST_MAILDOMAIN_LABEL="Blacklist mail domain"
```
Sort keys in alpha order. Capitalize label `Blacklist Mail Domain`.

Quy commented on language/en-GB/en-GB.com_users.ini:
```diff
@@ -126,6 +126,7 @@ COM_USERS_REGISTRATION_SAVE_FAILED="Registration failed: %s"
 COM_USERS_REGISTRATION_SAVE_SUCCESS="Thank you for registering. You may now log in using the username and password you registered with."
 COM_USERS_REGISTRATION_SEND_MAIL_FAILED="An error was encountered while sending the registration email. A message has been sent to the administrator of this site."
 COM_USERS_REGISTRATION_VERIFY_SUCCESS="Your email address has been verified. Once an administrator approves your account you will be notified by email and you can login to the site."
+COM_USERS_REGISTRATION_USER_MAIL_DOMAIN_NOT_ALLOWED_MESSAGE="Your mail domain <strong>%s</strong> is not allowed"
```
Sort key in alpha order. Add period.

carlitorweb committed:
Code review

carlitorweb committed:
TLD Support

carlitorweb commented:
Thank you again @Quy . Also, i added support for TLD.

tonypartridge commented:
Actually I support this for the core. Given Joomla’s user base at present this seems like a logical addition to the core.

brianteeman commented:
there really is no difference to the password limits except now we have some domain limits we already have

Quy commented on administrator/language/en-GB/en-GB.com_users.ini:
```diff
@@ -54,6 +56,8 @@ COM_USERS_CONFIG_FIELD_USERACTIVATION_DESC="If set to None the user will be regi
 COM_USERS_CONFIG_FIELD_USERACTIVATION_LABEL="New User Account Activation"
 COM_USERS_CONFIG_FIELD_USERACTIVATION_OPTION_ADMINACTIVATION="Administrator"
 COM_USERS_CONFIG_FIELD_USERACTIVATION_OPTION_SELFACTIVATION="Self"
+COM_USERS_CONFIG_FIELD_WHITELIST_MAILDOMAIN_LABEL="Whitelist Mail"
```
Sort in alpha order. Swap lines 59 & 60.

carlitorweb committed:
Sort in alpha order

brianteeman commented on language/en-GB/en-GB.com_users.ini:
```diff
@@ -125,6 +125,7 @@ COM_USERS_REGISTRATION_DEFAULT_LABEL="User Registration"
 COM_USERS_REGISTRATION_SAVE_FAILED="Registration failed: %s"
 COM_USERS_REGISTRATION_SAVE_SUCCESS="Thank you for registering. You may now log in using the username and password you registered with."
 COM_USERS_REGISTRATION_SEND_MAIL_FAILED="An error was encountered while sending the registration email. A message has been sent to the administrator of this site."
+COM_USERS_REGISTRATION_USER_MAIL_DOMAIN_NOT_ALLOWED_MESSAGE="Your mail domain <strong>%s</strong> is not allowed"
```
please change to "email domain"

carlitorweb committed:
Code review. Better explanation of the fields

infograf768 commented:
Tested here:

1. Creating user in back-end accepts blacklisted domains.
2. Modifying account in back-end also
3. Editing profile in front-end also

I therefore consider this proposal as incomplete.

carlitorweb commented:
Thank @infograf768 . What others change can be in this proposal?

carlitorweb committed:
Better label/description text

infograf768 commented:
To be acceptable, a validation should occur each time the email field is presented to be filled in a form.

dneukirchen commented:
First of all: thank you for the contribution!  I hate to say that, but at least for me this should not be part of core. Its a 3dp feature, which can easily be implemented using a plugin. We already have enough (confusing) edge case configuration settings.

carlitorweb commented:
> To be acceptable, a validation should occur each time the email field is presented to be filled in a form.

@infograf768 you mean, validate on the client side too?

> First of all: thank you for the contribution! I hate to say that, but at least for me this should not be part of core. Its a 3dp feature, which can easily be implemented using a plugin. We already have enough (confusing) edge case configuration settings.

You welcome @dneukirchen , I glad of contributing to my favorite tool. Hope I can do more. About your opinion, IMHO a lot of things inside Joomla can be as part of a 3dp feature, and not for that, need be out of Joomla. But no problem if this is the reason why this will not be considered to be added to Joomla.

Quy commented:
> you mean, validate on the client side too?

Here is his previous reply:
> Creating user in back-end accepts blacklisted domains.
> Modifying account in back-end also
> Editing profile in front-end also

In my case, I manage 10 Joomla sites that don't have user registrations so this feature is not applicable.

carlitorweb commented:
Thank you @Quy , but I still not understand what mean. Sorry for boring. 

> In my case, I manage 10 Joomla sites that don't have user registrations so this feature is not applicable.

Well, I decided to make this PR, after found 7 times this request. I read all the comments before about why this featured is not applicable, and is okay if this is a no need. How can I know this decision? Someone on the project come here and close this PR or I need close it?

Quy commented:
> Editing profile in front-end also

Go to the frontend.
Log in.
Edit your profile.
Change your email address to a blacklisted domain/tld which will be accepted when it should be rejected.

The same applies to the backend.

carlitorweb commented:
I got it now @Quy . Thank you, and sorry for no understand before

carlitorweb committed:
Validation creating user in the backend

carlitorweb committed:
Validation creating/editing user in the backend

stutteringp0et commented:
This functionality has existed as a plugin since 2011/J1.6.

carlitorweb committed:
Validation editing user in the frontend

carlitorweb committed:
Changing comment line

carlitorweb commented:
Added missing validation check when you create/edit a user on the backend. And added missing validation check when you edit a user on the frontend. And sorry again @Quy and @infograf768 for no understand well before when I left this checks out.

carlitorweb committed:
Changing comment line

infograf768 commented:
test:
issue solved when editing profile in frontend and editing user in back-end, but not when editing account in backend
link is `/administrator/index.php?option=com_admin&task=profile.edit&id=122` (where the id is the user id)
<img width="271" alt="screen shot 2018-05-02 at 09 34 27" src="https://user-images.githubusercontent.com/869724/39510979-0f87e8de-4dec-11e8-9cfd-cc8cfea38e86.png">
The model is `/administrator/components/com_admin/models/profile.php`

carlitorweb committed:
Validation editing account in backend

carlitorweb committed:
Validation editing account in backend

carlitorweb commented:
Thank you again @infograf768 . Is done

SharkyKZ commented:
Even if using two lists, you still need to add an additional option to define default action for domains which aren't on any list. Currently the logic is flawed. Test case:
Add blocked.com to blacklist.
Add allowed.com to whitelist.
Try to register with example.com
The registration will fail even though example.com is not explicitly blacklisted.

Still, I don't think it's correct to have two lists. Unless I'm forgetting some scenario, one list with wildcard support would be better.

carlitorweb commented:
But if you define something inside a whitelist, then you are defining the only patterns that are going to be allowed (is how work a whitelist imho). That is why **example.com** is not allowed. The condition check as first if something is inside of the blacklist and if it matches yours, if not, then go to the whitelist for doing same, but if the whitelist have some pattern inside, all different email domain/tld will fail, because the site only accept one or several specific patterns

infograf768 commented on administrator/components/com_admin/models/profile.php:
```diff
@@ -125,6 +125,26 @@ public function save($data)
 		$user = JFactory::getUser();
 
+		// Check if the user mail domain or TLD is disallowed
+		$config = JComponentHelper::getParams('com_users');
```
I suggest to use `$usersParams` instead of `$config` here (and change below).
This would fit the way it is used elsewhere in core.

carlitorweb commented:
Done. Thank you @infograf768

carlitorweb committed:
Code review

infograf768 commented:
If this PR is accepted, we may think later about adding a check also when entering an address in the mail field for a contact. Not sure if it is necessary.

In the mean while, I think this new feature is nice, simple and should get in.

SharkyKZ commented:
>But if you define something inside a whitelist, then you are defining the only patterns that are going to be allowed (is how work a whitelist imho).

This is true but the same is true for blacklist. Both lists are of equal importance. So, if there's no default option set by the user, you can't prioritize any of the lists in your code. And because of that, any domain that isn't on any list should be neither blocked nor allowed, which makes no sense. So you do need to add default action.

infograf768 commented:
Code should first check if whitelist contains anything. If it does, no need to use blacklist (else than sending an error if blacklist contains the same value as whitelist). If it is empty, check blacklist and act accordingly.

infograf768 commented:
@SharkyKZ @carlitorweb 
Afterthought: I think whitelist should have priority. If it is empty, then blacklist should take over.
If it is not empty, a check should be done to make sure the same entry does not exist in blacklist, throw an error if there is one or ignore blacklist if none.

carlitorweb commented:
> If this PR is accepted, we may think later about adding a check also when entering an address in the mail field for a contact. Not sure if it is necessary.

Good point. I think this field need have the same restriction.

> In the mean while, I think this new feature is nice, simple and should get in.

Thank you @infograf768 and for providing a test.

Follow all the opinions of @SharkyKZ, now you can put what restriction you want to use. As @SharkyKZ  said, really there is not a scenario where you need use both list. So, one is enough. 

![select-restriction](https://user-images.githubusercontent.com/3956472/39637924-76c43ad2-4f92-11e8-980f-b25a1082628c.gif)

I made a commit where you can test this when you create a user inside the backend. If is correct the way, I will make all the change to others place.

carlitorweb committed:
Option to select what restriction want use

carlitorweb committed:
Drone fix

carlitorweb committed:
Put new option adn code for frontend

carlitorweb committed:
New option and code for com_admin.profile

Fixed a miss check before see the whitelist

carlitorweb commented:
I update all the files, for include this new way. Look better now. Thank you @SharkyKZ @infograf768

Quy commented:
The same check/code are in 4 places. Can it be done with a single method?

carlitorweb commented:
For that, I think need an event inside plugin.user but I do not think this check/code it must be there. Not when the option is controlled by the component.

infograf768 commented on components/com_users/models/profile.php:
```diff
@@ -302,7 +302,38 @@ public function save($data)
 		// Prepare the data for the user object.
 		$data['email']    = JStringPunycode::emailToPunycode($data['email1']);
-		$data['password'] = $data['password1'];
+		$data['password'] = $data['password1']; 
```
Delete extraneous space after the ;

infograf768 commented:
The PR is now totally broken. Tested with blacklist domain.

Also: $params are missing when editing frontend profile
Also: I suggest the error lang string to be generic, i.e. added in en-GB.ini (front and backend) as 
``` 
JGLOBAL_USER_MAIL_DOMAIN_NOT_ALLOWED="This email domain <strong>%s</strong> is not allowed"
```

infograf768 commented:
As far as I can see isssue is that `$blackListMailDomain` as well as `$whiteListMailDomain` are not defined anymore.

infograf768 commented:
Tested this OK (Note the single quote for the `$optionRestriction` value):
```
			if (!empty(array_filter($listMailDomainTLD)))
			{
				if ($optionRestriction === '2' && !empty(array_intersect($needles, $listMailDomainTLD)))
				{
					$this->setError(JText::sprintf('COM_USERS_REGISTRATION_USER_MAIL_DOMAIN_NOT_ALLOWED_MESSAGE', $userMailDomain[1]));

					return false;
				}
				elseif ($optionRestriction === '1' && empty(array_intersect($needles, $listMailDomainTLD)))
				{
					$this->setError(JText::sprintf('COM_USERS_REGISTRATION_USER_MAIL_DOMAIN_NOT_ALLOWED_MESSAGE', $userMailDomain[1]));

					return false;
				}
			}
```

carlitorweb committed:
Fix missing variables. Fix conditions check.

carlitorweb committed:
Fix missing variables. Fix conditions checks

carlitorweb commented:
Thank you @infograf768 . Now is okay, and added the global string too.

carlitorweb committed:
Adding global string

infograf768 commented on administrator/language/en-GB/en-GB.com_admin.ini:
```diff
@@ -193,4 +193,4 @@ COM_USERS_MSG_NOT_ENOUGH_UPPERCASE_LETTERS_N_1="Password does not have enough up
 COM_USERS_MSG_PASSWORD_TOO_SHORT_N="Password is too short. Passwords must have at least %s characters."
 COM_USERS_MSG_SPACES_IN_PASSWORD="Password must not have spaces at the beginning or end."
-
+COM_USERS_MSG_USER_MAIL_DOMAIN_NOT_ALLOWED="Your mail domain <strong>%s</strong> is not allowed"
```
Can be deleted as we now use a global string

carlitorweb commented:
Thank you @infograf768  is done.

carlitorweb committed:
Deleting string com_admin

infograf768 commented:
I have tested this item :white_check_mark: successfully on 8882533336047ab7e2375b905a9bde29e9c64f91<br /><br />Looks OK now.

@SharkyKZ 

Please test again.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20164">issues.joomla.org/tracker/joomla-cms/20164</a>.</sub>

Quy commented:
I can still register with a blacklisted domain.

How about only show these settings when `Allow User Registration` is enabled?

carlitorweb committed:
Fix missing variables.

carlitorweb committed:
Show options only when User Registration is enabled

carlitorweb commented:
Good catch @Quy thank you. I swear I had updated that model with the new changes. Thank you. I test now 10 times, on the administration, and the frontend. Now i not have doubt all work, how need work.

And excellent suggestion, about only show these settings when Allow User Registration is enabled. Is done too.

carlitorweb committed:
Change string for the global one

Quy commented on administrator/components/com_users/models/user.php:
```diff
@@ -215,6 +215,38 @@ public function save($data)
+				}
+			}
+		}
+		
```
Remove tabs

Quy commented on administrator/language/en-GB/en-GB.com_users.ini:
```diff
@@ -21,6 +21,10 @@ COM_USERS_CONFIG_FIELD_CAPTCHA_DESC="Select the captcha plugin that will be used
 COM_USERS_CONFIG_FIELD_CAPTCHA_LABEL="Captcha"
 COM_USERS_CONFIG_FIELD_CHANGEUSERNAME_DESC="Allow users to change their Username when editing their profile."
 COM_USERS_CONFIG_FIELD_CHANGEUSERNAME_LABEL="Change Username"
+COM_USERS_CONFIG_FIELD_DOMAIN_TLD_RESTRICTION_DESC="Whitelist Email/TLD: Users can register only from the emails domains and/or TLD (Top level domain). <br /><br /> Blacklist Email/TLD: Users can register except those emails domains and TLD (Top level domain) in the blacklist. <br /><br /> Enter each domain or TLD in a new line. Example of a domain 'gmail.com' . Example of a TLD 'gov'"
```
Remove `in the blacklist` since there is only one text box for both whitelist/blacklist.

SharkyKZ commented:
After considering many options, I think it's best to implement robots.txt-like rules. This would allow for more precise control, including subdomains, wildcard support and user-set priority. And to prevent code duplication, this can be added as an option to EmailRule.

@carlitorweb do you mind if I submit an alternative PR?

carlitorweb commented:
@SharkyKZ  no problem.
